storage/queue.go

```diff
@@ -111,6 +112,96 @@ type PeekMessageResponse struct {
 	MessageText    string `xml:"MessageText"`
 }
 
+// QueueMetadata represents user defined metadata and queue properties on
+// a specific queue.
+//
+// See https://msdn.microsoft.com/en-us/library/azure/dd179384.aspx
+type QueueMetadata struct {
+	ApproximateMessageCount int
+	userDefinedMetadata     map[string]string
+}
+
+// Get retrieve a value for a case insensitive name.
+func (qm *QueueMetadata) Get(name string) (value string) {
+	qm.ensureInit()
+	value, _ = qm.userDefinedMetadata[strings.ToUpper(name)]
+	return
+}
+
+// Set sets a name/value metadata pair.
+func (qm *QueueMetadata) Set(name string, value string) {
+	qm.ensureInit()
+	qm.userDefinedMetadata[strings.ToUpper(name)] = value
+	return
+}
+
+// Del remove a metadata entry.
+func (qm *QueueMetadata) Del(name string) {
+	qm.ensureInit()
+	delete(qm.userDefinedMetadata, strings.ToUpper(name))
+}
+
+func (qm *QueueMetadata) ensureInit() {
+	if qm.userDefinedMetadata == nil {
+		qm.userDefinedMetadata = make(map[string]string)
+	}
+}
+
+// SetMetadata operation sets user-defined metadata on the specified queue.
+// Metadata is associated with the queue as name-value pairs.
+//
+// See https://msdn.microsoft.com/en-us/library/azure/dd179384.aspx
+func (c QueueServiceClient) SetMetadata(name string, qm QueueMetadata) error {
+	uri := c.client.getEndpoint(queueServiceName, pathForQueue(name), url.Values{"comp": []string{"metadata"}})
+	headers := c.client.getStandardHeaders()
+	headers["Content-Length"] = "0"
+	for k, v := range qm.userDefinedMetadata {
+		headers[userDefinedMetadataHeaderPrefix+k] = v
```

Comment:
That's about where it currently stands. (the type name is QueueMetadataResponse, since it is namespaced under storage. and is only relevant to Queue)

As for PATCH semantics - this azure API do not support it. the PUT request sets whatever is sent and discards all the rest (which is proper for PUT).